pkg/didmethod/peer/did.go

```diff
@@ -30,7 +31,77 @@ func newDid(doc *did.Doc) (string, error) {
 	if doc.PublicKey == nil || doc.Authentication == nil {
 		return "", errors.New("the genesis version must include public keys and authentication")
 	}
+	numBasis, err := numBasis(doc)
+	if err != nil {
+		return "", err
+	}
+	messageIdentifier := []string{peerPrefix, numAlgo, encAlgo, "-", numBasis}
+	peerDID := strings.Join(messageIdentifier, "")
+
+	return peerDID, nil
+}
+
+/*
+  validate function validates the peer DID doc/resolved variant (doc which would have DID value in the root id property):
+  1. Calculate numbasis of the genesis version (without DID) of the peer DID doc
+  2. Compare the calculated numbasis in the step 1 with the extracted numbasis part of the DID value
+
+  Reference : https://openssi.github.io/peer-did-method-spec/index.html#create-register
+*/
+func validate(doc *did.Doc) error {
+
+	peerDid := doc.ID
+
+	err := validateDid(peerDid)
+	if err != nil {
+		return err
+	}
 
+	extractedNumBasis := extractNumBasis(peerDid)
+
+	// genesis version(no did) of the peer DID doc
+	genesisDoc := &did.Doc{
+		Context:        doc.Context,
+		ID:             "",
+		PublicKey:      doc.PublicKey,
+		Service:        doc.Service,
+		Authentication: doc.Authentication,
+		Created:        doc.Created,
+		Updated:        doc.Updated,
+		Proof:          doc.Proof}
+
+	//calculate the numbasis of the genesis version of the peer DID doc
+	numBasis, err := numBasis(genesisDoc)
+	if err != nil {
+		return err
+	}
+
+	if !(numBasis == extractedNumBasis) {
+		return errors.New("hash of the doc doesnt match the computed hash")
+	}
+	return nil
+}
+
+/* validateDid validates the did using the regular expression mentioned in the spec.
+   Reference: https://dhh1128.github.io/peer-did-method-spec/#matching-regex */
+func validateDid(did string) error {
+	matched, _ := regexp.MatchString(`did:peer:11-([a-fA-F0-9]){64}`, did)
+	if !matched {
+		return errors.Errorf("did doesnt follow matching regex")
+	}
+	return nil
+}
+
+// extractNumbasis extract the portion that returns the numBasis of a given peer DID
+func extractNumBasis(peerDid string) string {
+	splitDid := strings.FieldsFunc(peerDid, func(r rune) bool { return r == '-' })
+	numBasis := splitDid[1]
```

Comment:
Because this may result in a panic if the input argument is malformed, I suggest we remove this function and inline the splitting where it's needed (right after the format is validated).